cpp/unreal_plugins/SpEngine/Source/SpEngine/WorkQueue.h

```diff
@@ -0,0 +1,133 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <concepts>    // std::same_as
+#include <cmath>       // std::nan
+#include <future>
+#include <map>
+#include <mutex>
+#include <string>
+#include <utility>     // std::forward, std::move
+#include <vector>
+
+#include <SpCore/Assert.h>
+#include <SpCore/Boost.h>
+
+template <typename TFunc, typename TReturn, typename... TArgs>
+concept CCallable = requires(TFunc func, TArgs... args) {
+        { func(args...) } -> std::same_as<TReturn>;
+};
+
+
+template <typename TClass>
+struct FuncInfo : public FuncInfo<decltype(&TClass::operator())> {};
+
+template <typename TClass, typename TReturn, typename... TArgs>
+struct FuncInfo<TReturn(TClass::*)(TArgs...)> : public FuncInfo<TReturn(*)(TArgs...)> {};
+
+template <typename TClass, typename TReturn, typename... TArgs>
+struct FuncInfo<TReturn(TClass::*)(TArgs...) const> : public FuncInfo<TReturn(*)(TArgs...)> {};
+
+template <typename TReturn, typename... TArgs>
+struct FuncInfo<TReturn(*)(TArgs...)> {};
+
+
+// The purpose of this class is to provide a copy-constructible type that derives from std::packaged_task. This is necessary 
+// because boost::asio requires that task objects are copy-constructible, but std::packaged_task is not copy-constructible.
+template <typename TReturn>
+struct CopyConstructibleTask : std::packaged_task<TReturn()>
+{
+    CopyConstructibleTask(auto&& func) : std::packaged_task<TReturn()>(std::forward<decltype(func)>(func)) {};
+};
+
+
+class WorkQueue {
```

Comment:
Organize this file so high-level functions come first and then low-level functions come after, so people can read it from top-to-bottom.

```
// concepts

class WorkQueue
{
public:
    // high-level functions
    // low-level functions

private:
    // FuncInfo
    // private data members
}

Reply:
Done